src/Processors/QueryPlan/Optimizations/filterPushDown.cpp

```diff
@@ -428,6 +428,9 @@ size_t tryPushDownFilter(QueryPlan::Node * parent_node, QueryPlan::Nodes & nodes
         /// of the grouping sets, we could not push the filter down.
         if (aggregating->isGroupingSets())
         {
+            /// Cannot push down filter if type has been changed.
+            if (aggregating->isGroupByUseNulls())
+                return 0;
```

Comment:
Makes sense. But it sad, that `group_by_use_nulls` limits optimizations so much..Do you think it's possible to update the types in the DAG?

Reply:
Probably yes, but I cannot come up with solution from the top of my head